src/nouveau/src/nouveau_fabric_search.erl

```diff
@@ -0,0 +1,213 @@
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+%% -*- erlang-indent-level: 4;indent-tabs-mode: nil -*-
+
+-module(nouveau_fabric_search).
+
+-export([go/4]).
+
+-include_lib("mem3/include/mem3.hrl").
+-include_lib("couch/include/couch_db.hrl").
+
+-record(state, {
+    limit,
+    sort,
+    counters,
+    search_results
+}).
+
+go(DbName, GroupId, IndexName, QueryArgs0) when is_binary(GroupId) ->
+    {ok, DDoc} = fabric:open_doc(
+        DbName,
+        <<"_design/", GroupId/binary>>,
+        [ejson_body]
+    ),
+    go(DbName, DDoc, IndexName, QueryArgs0);
+go(DbName, #doc{} = DDoc, IndexName, QueryArgs0) ->
+    {ok, Index} = nouveau_util:design_doc_to_index(DbName, DDoc, IndexName),
+    Shards = mem3:shards(DbName),
+    Bookmark = nouveau_bookmark:unpack(DbName, maps:get(bookmark, QueryArgs0)),
+    QueryArgs1 = maps:remove(bookmark, QueryArgs0),
+    Counters0 = lists:map(
+        fun(#shard{} = Shard) ->
+            After = maps:get(Shard#shard.range, Bookmark, null),
+            Ref = rexi:cast(
+                Shard#shard.node,
+                {nouveau_rpc, search, [Shard#shard.name, Index, QueryArgs1#{'after' => After}]}
+            ),
+            Shard#shard{ref = Ref}
+        end,
+        Shards
+    ),
+    Counters = fabric_dict:init(Counters0, nil),
+    Workers = fabric_dict:fetch_keys(Counters),
+    RexiMon = fabric_util:create_monitors(Workers),
+    State = #state{
+        limit = maps:get(limit, QueryArgs0),
```

Comment:
A minor nit: It's a bit odd to use a `QueryArgs0` after we've updated it to `QueryArgs1`. May be better to extract `Limit` and `Sort` initially and keep them as separate variables with `#{limit = Limit, sort = Sort} = QueryArgs0` somewhere before we updated to `QueryArgs1`